src/core/job.c

```diff
@@ -1173,7 +1178,32 @@ void job_add_to_run_queue(Job *j) {
                         log_warning_errno(r, "Failed to enable job run queue event source, ignoring: %m");
         }
 
-        LIST_PREPEND(run_queue, j->manager->run_queue, j);
+        /* Deterministic run-queue order:
+         * - mount and service units after any other units
+         * - units with higher CPU weight first
+         * - sort by id if the other criteria are the same */
+        state = manager_state(j->manager);
+        c = unit_get_cgroup_context(j->unit);
+        weight = c ? cgroup_context_cpu_weight(c, state) : CGROUP_WEIGHT_DEFAULT;
+        LIST_FOREACH(run_queue, tmp, j->manager->run_queue) {
+                if (!IN_SET(j->unit->type, UNIT_SERVICE, UNIT_MOUNT) &&
+                    IN_SET(tmp->unit->type, UNIT_SERVICE, UNIT_MOUNT))
+                        break;
+                if (IN_SET(j->unit->type, UNIT_SERVICE, UNIT_MOUNT) &&
+                    !IN_SET(tmp->unit->type, UNIT_SERVICE, UNIT_MOUNT))
+                        continue;
+                if (IN_SET(j->unit->type, UNIT_SERVICE, UNIT_MOUNT)) {
+                        c = unit_get_cgroup_context(tmp->unit);
+                        weight_other = c ? cgroup_context_cpu_weight(c, state) : CGROUP_WEIGHT_DEFAULT;
+                        if (weight > weight_other)
+                                break;
+                        if (weight < weight_other)
+                                continue;
```

Comment:
We have the `CMP` macro. Maybe it'd make the code a bit shorter here?

Reply:
I moved this elsewhere, but it uses `CMP` now.